File: android/lib/src/main/java/com/swmansion/gesturehandler/GestureUtils.java
package com.swmansion.gesturehandler;

import android.view.MotionEvent;

public class GestureUtils {
    public static float getLastPointerX(GestureHandlerMotionEventAdapter event, boolean averageTouches) {
        float offset = event.getRawX() - event.getX();
        int excludeIndex = event.getActionMasked() == MotionEvent.ACTION_POINTER_UP ?
                event.getActionIndex() : -1;

        if (averageTouches) {
            float sum = 0f;
            int count = 0;
            for (int i = 0, size = event.getMotionEventPointerCount(); i < size; i++) {
                if (i != excludeIndex && event.containsIndexOfMotionEvent(i)) {
                    sum += event.getX(i) + offset;
                    count++;
                }
            }
            return sum / count;
        } else {
            int lastPointerIdx = event.getMotionEventPointerCount() - 1;
            while (lastPointerIdx == excludeIndex || !event.containsIndexOfMotionEvent(lastPointerIdx)) {
                lastPointerIdx--;
            }
            return event.getX(lastPointerIdx) + offset;
        }
    }

    public static float getLastPointerY(GestureHandlerMotionEventAdapter event, boolean averageTouches) {
        float offset = event.getRawY() - event.getY();
        int excludeIndex = event.getActionMasked() == MotionEvent.ACTION_POINTER_UP ?
                event.getActionIndex() : -1;

        if (averageTouches) {
            float sum = 0f;
            int count = 0;
            for (int i = 0, size = event.getMotionEventPointerCount(); i < size; i++) {
                if (i != excludeIndex && event.containsIndexOfMotionEvent(i)) {
                    sum += event.getY(i) + offset;
                    count++;
                }
            }
            return sum / count;
        } else {
            int lastPointerIdx = event.getMotionEventPointerCount() - 1;
            while (lastPointerIdx == excludeIndex || !event.containsIndexOfMotionEvent(lastPointerIdx)) {
                lastPointerIdx--;
            }
            return event.getY(lastPointerIdx) + offset;
        }
    }
}

File: android/lib/src/main/java/com/swmansion/gesturehandler/GestureUtils.kt
package com.swmansion.gesturehandler

import android.view.MotionEvent

object GestureUtils {
  fun getLastPointerX(event: MotionEvent, averageTouches: Boolean): Float {
    val offset = event.rawX - event.x
    val excludeIndex = if (event.actionMasked == MotionEvent.ACTION_POINTER_UP) event.actionIndex else -1
    return if (averageTouches) {
      var sum = 0f
      var count = 0
      var i = 0
      val size = event.pointerCount
      while (i < size) {
        if (i != excludeIndex) {
          sum += event.getX(i) + offset
          count++
        }
        i++
      }
      sum / count
    } else {
      var lastPointerIdx = event.pointerCount - 1
      if (lastPointerIdx == excludeIndex) {
        lastPointerIdx--
      }
      event.getX(lastPointerIdx) + offset
    }
  }

  fun getLastPointerY(event: MotionEvent, averageTouches: Boolean): Float {
    val offset = event.rawY - event.y
    val excludeIndex = if (event.actionMasked == MotionEvent.ACTION_POINTER_UP) event.actionIndex else -1
    return if (averageTouches) {
      var sum = 0f
      var count = 0
      var i = 0
      val size = event.pointerCount
      while (i < size) {
        if (i != excludeIndex) {
          sum += event.getY(i) + offset
          count++
        }
        i++
      }
      sum / count
    } else {
      var lastPointerIdx = event.pointerCount - 1
      if (lastPointerIdx == excludeIndex) {
        lastPointerIdx -= 1
      }
      event.getY(lastPointerIdx) + offset
    }
  }
}